### ISSUE TITLE
Gate_output is being mis-predicted

### ISSUE THREAD

**Issue by tdplaza** [NVIDIA/tacotron2#319]
I have trained the model for only few steps (8000) for testing.

At this checkpoint, the synthesized mel spectrogram's frames always bigger than max frames  (in hparams config)

As I have checked, the problem is gate_output always less than configured gate_threshold (it's around 0.0022 or 0.0023), so the frames had been generated forever.

Is this problem because I was not trained model much enough, or what should I have to check? 

Please help, thank you so much.

**Comment by CookiePPP** [comment]
1. **What does the alignment look like?**
Are you doing 8000 steps from a pretrained model or from scratch?
If the model has not aligned to the text then the model will not be able to predict when to finish synthesis.

2. **Do your audio files end when the speaker stops speaking or is there some silences at the end?**
This may give the model trouble predicting the gate_output.
(though silence at the end is normally fine as long as it doesn't go on for more than a second)

3. You can change
https://github.com/NVIDIA/tacotron2/blob/master/loss_function.py#L18
```
        gate_loss = nn.BCEWithLogitsLoss()(gate_out, gate_target)
```
to
```
        gate_loss = nn.BCEWithLogitsLoss(pos_weight=20)(gate_out, gate_target)
```
(Change 20 to whatever you want, though 20 has served me well. Default for BCEWithLogitsLoss is 1)
This will balance out the effect of having the model only predict a single positive gate_out but many 0's during training by giving the positive a higher weight.

**Comment by tdplaza** [comment]
@CookiePPP Thank your for your answer.

I trained model from scratch, because the language is different from English.

My training audios have silence after speaker voice.

I will check the alignment and loss function as your suggestion. Thank you so much.

**Comment by CookiePPP** [comment]
@tdplaza 
You will likely need to train for longer.

**Comment by tdplaza** [comment]
@CookiePPP 
I have train more step (to 20,000), but the alignment seems to be not going to aligned, nothing changed. 

I know this is hard to support, just want to ask you, how much steps did you train to get it aligned? Of course, any suggestion is appreciated. Thank you.

![Figure_1](https://user-images.githubusercontent.com/31118781/77605658-d9cb3580-6f58-11ea-957d-1a3f3e71f51a.png)

**Comment by CookiePPP** [comment]
@tdplaza 
Do you have tensorboard set up?

-------------------------

## Training
1. `python train.py --output_directory=outdir --log_directory=logdir`
2. (OPTIONAL) `tensorboard --logdir=outdir/logdir`

------------------------

Tensorboard images tab shows alignments a little better.

**Comment by tdplaza** [comment]
This is my tensorboard output

![image](https://user-images.githubusercontent.com/31118781/77606866-0df42580-6f5c-11ea-9687-c42831b3aa70.png)
![image](https://user-images.githubusercontent.com/31118781/77606898-1a787e00-6f5c-11ea-8cc1-5404893a2f2c.png)

**Comment by CookiePPP** [comment]
@tdplaza 
Images tab? (along the top)

**Comment by tdplaza** [comment]
![image](https://user-images.githubusercontent.com/31118781/77607480-ce2e3d80-6f5d-11ea-8b91-fc3cf336da1a.png)
![image](https://user-images.githubusercontent.com/31118781/77607497-dd14f000-6f5d-11ea-8b81-03e0cbc7d44b.png)
![image](https://user-images.githubusercontent.com/31118781/77607502-e605c180-6f5d-11ea-8826-e0e5aee38358.png)

I saw the bad result, but dont now where to check and fix. Thank you for your patience supports

**Comment by CookiePPP** [comment]
It looks wrong but I don't know why.
Your audio files look very quiet.
Also `mel_predicted` should look like `mel_target`, even without alignment they should look closer.

Check your audio files are loud enough and are being read properly.

**Comment by tdplaza** [comment]
Thank you so much, I will find it out.

**Comment by tdplaza** [comment]
Dear @CookiePPP,

I have solved the mel-spectrogram problem, but at 48k traning steps, the alignment is as following: 
![image](https://user-images.githubusercontent.com/31118781/77864797-c3c5b980-7265-11ea-9368-50e20194c40d.png)

And the predicted mel is: 
![image](https://user-images.githubusercontent.com/31118781/77864815-dfc95b00-7265-11ea-8df9-1aa47e06dc5a.png)

Could you please tell me, is the alignment is normal? I mean, with your experiments, at 48k steps, is above alignment is normal? I don't know whether it will be aligned (with more training steps) or not.
My Tensorflow version of Tacotron-2 ([here](https://github.com/tdplaza/Tacotron_Prosody_multi_GPU)) can pretty well aligned at ~30k steps.

How many steps did you train to get good result? both Tacotron and WaveGlow.

And, from [here](https://github.com/NVIDIA/tacotron2/issues/321#issuecomment-604194446), I saw you recommend using pretrained WaveGlow for training different language, is it really possible?

Thank you so much for your kindness, sir.

**Comment by CookiePPP** [comment]
@tdplaza 
[This gif](https://user-images.githubusercontent.com/39617376/68069699-ddd69b00-fd9e-11e9-8d32-eec15e84eb30.gif) shows this Tacotron2 repo should align after about 25,000 steps for English.
(https://github.com/NVIDIA/tacotron2/issues/280)

I don't know if it will align with more steps or not.

------------------------------

> pretrained WaveGlow for training different language, is it really possible?

I'd have to test it, but it should produce okay audio.
Human's all use the same vocal cords etc, just slightly different sounds.

**Comment by tdplaza** [comment]
Thank you so much,

Since my batch_size is  just 12, I will train more and see what will happen.

**Comment by EuphoriaCelestial** [comment]
> For more information, I'm using GTX 1080 Ti (which does not have Tensor Cores to support Mixed Precision training, but for saving memory, I'm training with fp16_run=True and batch_size=12)

I have 2080ti which could train with fp16=False and batch_size=32. I would like to know if fp16 affect the accuracy of training result much? because I want to increase batch_size a little bit more

**Comment by EuphoriaCelestial** [comment]
@tdplaza can I have your email to ask something about data pre-processing? I believe you and I training on same language :)

**Comment by tdplaza** [comment]
@EuphoriaCelestial 

As the Mixed Precision's authors claimed, [the paper](https://arxiv.org/abs/1710.03740)

Using `fp16_run=True` is as good as fp32, of course, training with mixed precision requires Tensor cores GPUs





> I believe you and I training on same language :)

I don't think so, my email is thien.tdplaza@gmail.com

**Comment by Jeevesh8** [comment]
@tdplaza How did you solve that [mel-spec problem](https://github.com/NVIDIA/tacotron2/issues/319#issuecomment-605724973) ?

**Comment by tdplaza** [comment]
@Jeevesh8 Which proplem are you mentioning here?

The first problem was the mel target is almost empty. Because my corpus audio's data scale was float32, But this repo works with int16 data scale. Then I re-normalized them to int16 before processing.

The second problem was this repo misbehaved when training with audios that has silent part at beginning and ending, then I cut them out.

The third is the last character of sentence. When training, you trained with "." at the end of every sentence, then add "." (not " ." or ". ") at the end when synthesizing.

and the last is you need enough training. For batchsize 64, corpus with 12 hours of single speaker speech, at 20000 ~ 30000 training steps attention should be aligned.

**Comment by Jeevesh8** [comment]
Thank you very much @tdplaza  for your kindness.

**Comment by indexalice** [comment]
@tdplaza Sorry for bothering.
Did you do any more preprocessing on your own dataset? I have similar images and I am confused about **the blank green space in predicted mel and target mel**. Is it normal? Will it result in no alignment? Any response will be appreciated.
![屏幕截图 2020-06-26 16 22 23](https://user-images.githubusercontent.com/20719001/85831693-a0a99800-b7c9-11ea-962d-f9f3335271ad.png)
![屏幕截图 2020-06-26 16 23 50](https://user-images.githubusercontent.com/20719001/85831702-a56e4c00-b7c9-11ea-9620-c8adde3e2918.png)

**Comment by EuphoriaCelestial** [comment]
@indexalice you only need to keep training for longer time

**Comment by indexalice** [comment]
@EuphoriaCelestial Thank you for your kind response! I will keep training and update the result later.

**Comment by indexalice** [comment]
@EuphoriaCelestial 
After 300 epochs, the alignment result is still the same. My own dataset for fine-tuning is about 3 hours (It is a little small I see) and I reduce the learning_rate to 2e5. Or should I use a larger learning rate?

By the way, may I ask you some questions about fine-tuning Tacotron by email? 
![屏幕截图 2020-06-28 19 17 02](https://user-images.githubusercontent.com/20719001/85945721-aa263200-b97a-11ea-868d-8d01bb2d6632.png)
![屏幕截图 2020-06-28 19 17 59](https://user-images.githubusercontent.com/20719001/85945724-adb9b900-b97a-11ea-96a8-a67ddfd67fff.png)

**Comment by Jeevesh8** [comment]
@indexalice  please ask here only, I am having similar problems.

**Comment by EuphoriaCelestial** [comment]
@indexalice with only 3 hours dataset and such small learning rate, 300 epochs mean nothing. I suggest you increase learning rate to 1e-3 or 5e-4, keep training until it get around 1000 epochs (it shouldnt take long since your dataset is too small)
I know little to nothing about fine tuning, so I dont think I can help

**Comment by indexalice** [comment]
@EuphoriaCelestial Thank you very much for the suggestion! I will try it out.
